xds/googledirectpath/googlec2p.go

```diff
@@ -0,0 +1,219 @@
+
+	logPrefix = "[google-c2p-resolver]"
+)
+
+// For overriding in unittests.
+var (
+	onGCE = func() bool {
+		return googlecloud.OnGCE()
+	}
+	newClientWithConfig = func(config *bootstrap.Config) error {
+		_, err := xdsclient.NewWithConfig(config)
+		return err
+	}
+
+	childBuilders = map[string]resolver.Builder{
+		"dns": resolver.Get("dns"),
+		"xds": resolver.Get("xds"),
+	}
+
+	logger = internalgrpclog.NewPrefixLogger(grpclog.Component("directpath"), logPrefix)
+
+	defaultNode = &v3corepb.Node{
+		Id:                   "C2P",
+		Metadata:             nil, // To be set if ipv6 is enabled.
+		Locality:             nil, // To be set to the value from metadata.
+		UserAgentName:        gRPCUserAgentName,
+		UserAgentVersionType: &v3corepb.Node_UserAgentVersion{UserAgentVersion: grpc.Version},
+		ClientFeatures:       []string{clientFeatureNoOverprovisioning},
+	}
+
+	ipv6EnabledMetadata = &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			ipv6CapableMetadataName: {
+				Kind: &structpb.Value_BoolValue{BoolValue: true},
+			},
+		},
+	}
+)
+
+func init() {
+	resolver.Register(&c2pResolverBuilder{})
+}
+
+type c2pResolverBuilder struct{}
+
+// Build helps implement the resolver.Builder interface.
+func (b *c2pResolverBuilder) Build(t resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	ret := &c2pResolver{
+		target: t,
+		cc:     cc,
+		opts:   opts,
+	}
+	// start does I/O. Run it in a goroutine.
+	go ret.start()
+	return ret, nil
+}
+
+// Name helps implement the resolver.Builder interface.
+func (*c2pResolverBuilder) Scheme() string {
+	return c2pScheme
+}
+
+type c2pResolver struct {
+	target resolver.Target
+	cc     resolver.ClientConn
+	opts   resolver.BuildOptions
+
+	mu    sync.Mutex
+	done  bool
+	child resolver.Resolver
+}
+
+func (r *c2pResolver) start() {
+	if !runDirectPath() {
+		r.startChild("dns")
+		return
+	}
+
+	nodeCopy := proto.Clone(defaultNode).(*v3corepb.Node)
+	nodeCopy.Locality = &v3corepb.Locality{Zone: getZone()}
+	if getIPv6Capable() {
+		nodeCopy.Metadata = ipv6EnabledMetadata
+	}
+	config := &bootstrap.Config{
+		BalancerName: tdURL,
+		Creds:        grpc.WithCredentialsBundle(google.NewDefaultCredentials()),
+		TransportAPI: version.TransportV3,
+		NodeProto:    nodeCopy,
+	}
+
+	// Create singleton xds client with this config. The xds client will be
+	// used by the xds resolver later.
+	err := newClientWithConfig(config)
+	if err != nil {
+		r.cc.ReportError(fmt.Errorf("failed to start xDS client: %v", err))
+		return
+	}
+
+	r.startChild("xds")
+}
+
+func (r *c2pResolver) startChild(scheme string) {
+	t := r.target
+	t.Scheme = scheme
+
+	b, ok := childBuilders[scheme]
+	if !ok {
+		logger.Errorf("unknown child scheme: %q", scheme)
+		r.cc.ReportError(fmt.Errorf("unknown child scheme in c2p resolver: %q", scheme))
+		return
+	}
+	child, err := b.Build(t, r.cc, r.opts)
+	if err != nil {
+		r.cc.ReportError(err)
+		return
+	}
```

Comment:
This is interesting.  In this state we shouldn't have even allowed initialization to complete, and we're stuck forever.  Are we sure this should be asynchronous?

Reply:
This error is in `Build()` now